### ISSUE TITLE
Cross-Origin Request Blocked

### ISSUE THREAD

**Issue by billd100** [grpc/grpc-web#722]
I am using a React front-end with Envoy and a Go server. When I attempt to call the gRPC service from the front-end, the request gets blocked.

I've tried adding `{"Access-Control-Allow-Origin": "*"}` to the request metadata and have also tried disabling the pre-flight check when I get an instance of the GreeterClient but neither has worked. Can anyone advise on what I might try next?

Here is the React component that calls the servie:
```
import React from 'react';
import logo from './logo.svg';
import './App.css';


const { GreeterClient } = require('./helloworld_grpc_web_pb');
const { HelloRequest, HelloResponse } = require('./helloworld_pb.js');


var client = new GreeterClient('http://localhost:8080');


class Dashboard extends React.Component {

    callGrpcService = () => {
        const request = new HelloRequest();
        request.setName('Hello World!');


        client.sayHello(request, {"Access-Control-Allow-Origin": "*"}, (err, response) => {
            if (response == null) {
                console.log(err)
            } else {
                console.log(response)
            }
        });
    }


    render() {
        return (
            <div className="App">
                <header className="App-header">
                    <img src={logo} className="App-logo" alt="logo" />
                    <button style={{ padding: 10 }} onClick={this.callGrpcService}>Click for grpc request</button>
                </header>
            </div>
        );
    }
}


export default Dashboard;
```

envoy.yaml:
```
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address: { address: 0.0.0.0, port_value: 9901 }

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address: { address: 0.0.0.0, port_value: 8080 }
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route:
                  cluster: helloworld_service
                  max_grpc_timeout: 0s
              cors:
                allow_origin_string_match:
                - prefix: "*"
                allow_methods: GET, PUT, DELETE, POST, OPTIONS
                allow_headers: keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,custom-header-1,x-accept-content-transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web,grpc-timeout
                max_age: "1728000"
                expose_headers: custom-header-1,grpc-status,grpc-message
          http_filters:
          - name: envoy.grpc_web
          - name: envoy.cors
          - name: envoy.router
  clusters:
  - name: helloworld_service
    connect_timeout: 0.25s
    type: logical_dns
    http2_protocol_options: {}
    lb_policy: round_robin
    hosts: [{ socket_address: { address: localhost, port_value: 9090 }}]
```

The error in the console is:
```Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:8080/helloworld.Greeter/SayHello. (Reason: CORS request did not succeed).
2
```
And also:
```
Object { code: 2, message: "Http response at 400 or 500 level", metadata: undefined }```

**Comment by stanley-cheung** [comment]
You don't need to add this `{"Access-Control-Allow-Origin": "*"}` to the metadata argument.

Can you open up dev console, the Networks tab, to see if there's an OPTIONS request (that's the CORS preflight request sent by the browser), and what are the details there? 

What you had seems to be the same as our provided example - so it should work. As it stands, it seems that your Envoy instance is not responding to the OPTIONS request properly

**Comment by billd100** [comment]
Thanks for taking the time to respond, @stanley-cheung.

The OPTIONS request does get generated but it never seems to return. Below is what I see in the console. I am also confused that I am running the exact example.

![image](https://user-images.githubusercontent.com/9690096/73802037-4f512b80-478a-11ea-8dd3-144faa529be4.png)

**Comment by stanley-cheung** [comment]
Yea so your OPTIONS request never get responded to properly. I suppose your envoy instance is running at localhost:10000? I am not sure why that's the case. It works for me. This is the [envoy.yaml](https://github.com/grpc/grpc-web/blob/master/net/grpc/gateway/examples/echo/envoy.yaml) that I have been testing with. There had been some updates on the syntax on the envoy.yaml file recently but looks like you get the latest copy.

**Comment by kevinmichaelchen** [comment]
No need to reopen this issue, but for anyone else out there who runs into the issue where the OPTIONS request succeeds but the following POST request fails, I was running into this issue because the POST request was sending an `Authorization` header and I forgot to add that to `cors.allow_headers` in `envoy.yaml`.

**Comment by jonathanou** [comment]
I'm running into this exact issue, @billd100  were you able to solve this issue?

In Firefox and Chrome, when OPTIONS is made, the response doesn't have `access-control-allow-origin`:
![image](https://user-images.githubusercontent.com/1291962/103728124-15327e80-5018-11eb-9008-45728aa97e6f.png)

Strangely, in Microsoft Edge browser, it works.  It appears Edge never sends OPTIONS, but directly sends POST:

![image](https://user-images.githubusercontent.com/1291962/103728010-b7059b80-5017-11eb-8263-ce4ffa74bb6c.png)

I am using the latest `envoy.yaml` file as the hello world example one is using.  Is Envoy misconfigured somehow so that it isn't adding `access-control-allow-origin` to the response to OPTIONS?

**Comment by jonathanou** [comment]
To answer my own question, turns out I had another application starting an embedded web server on the same port 8080 on ipv4, so Firefox and Chrome was hitting the embedded web server and not my Envoy proxy!

But why does Edge browser work?  Turns out Edge browser defaults to using ipv6 for `localhost`, and the embedded web server was only listening on 8080 using ipv4, not ipv6.  Envoy proxy listens to both ipv4 and ipv6, which is why Edge was able to make the request using ipv6.

As for why In Edge I was not seeing OPTIONS being sent, this is because the response of OPTIONS is cached once it is determined CORS is allowed.  In Firefox, after I switched Envoy to listen at a different port, the subsequent OPTIONS request is no longer made too.

One puzzling behavior is how it is possible for multiple processes to listen on the same port without any errors, apparently in Windows the behavior is [the second process will steal the port from the first process](https://stackoverflow.com/questions/17212789/multiple-processes-listening-on-the-same-port).

Bottom line is using 8080 as the port to listen for the hello world example probably should be changed to something less common.  On machines with lots of applications installed, there's a high likelihood there is someone listening on 8080.  This port stealing behavior is especially surreptitious and is really difficult to figure out.

**Comment by paymog** [comment]
What do we do if no `OPTIONS` request is being sent but we're still running into this CORS issue?

Here are the requests being sent by my browser, Brave:

![image](https://user-images.githubusercontent.com/1986950/118171946-d1061f00-b3f9-11eb-915c-bf26fc3c15fc.png)

Here's the full error I'm seeing in my console:

```
Access to XMLHttpRequest at 'localhost:8080/pipeline.Web/HelloWorld' from origin 'http://localhost:3000' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, brave, chrome-untrusted, https.
```

Here's my envoy config

```
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address: { address: 0.0.0.0, port_value: 9901 }

static_resources:
  listeners:
    - name: listener_0
      address:
        socket_address: { address: 0.0.0.0, port_value: 8080 }
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                codec_type: auto
                stat_prefix: ingress_http
                route_config:
                  name: local_route
                  virtual_hosts:
                    - name: local_service
                      domains: ["*"]
                      routes:
                        - match: { prefix: "/" }
                          route:
                            cluster: echo_service
                            timeout: 0s
                            max_stream_duration:
                              grpc_timeout_header_max: 0s
                      cors:
                        allow_origin_string_match:
                          - prefix: "*"
                        allow_methods: GET, PUT, DELETE, POST, OPTIONS
                        allow_headers: keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,custom-header-1,x-accept-content-transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web,grpc-timeout
                        max_age: "1728000"
                        expose_headers: custom-header-1,grpc-status,grpc-message
                http_filters:
                  - name: envoy.filters.http.grpc_web
                  - name: envoy.filters.http.cors
                  - name: envoy.filters.http.router
  clusters:
    - name: echo_service
      connect_timeout: 0.25s
      type: logical_dns
      http2_protocol_options: {}
      lb_policy: round_robin
      load_assignment:
        cluster_name: cluster_0
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 6789
```

Note that I changed the address of the cluster endpoint (the 2nd last line) to be `127.0.0.1` for local testing. Also note that my GRPC server is running on port 6789 which has been reflected in the envoy config.
